lib/elixir/lib/calendar/iso.ex

```diff
@@ -339,7 +339,128 @@ defmodule Calendar.ISO do
   @impl true
   def day_of_week(year, month, day)
       when is_integer(year) and is_integer(month) and is_integer(day) do
-    Integer.mod(date_to_iso_days(year, month, day) + 5, 7) + 1
+    iso_days_to_day_of_week(date_to_iso_days(year, month, day))
+  end
+
+  defp iso_days_to_day_of_week(iso_days) do
+    Integer.mod(iso_days + 5, 7) + 1
+  end
+
+  @doc """
+  Returns the week of a given date according to ISO.
+
+      # First day of the year is a monday
+      iex> Calendar.ISO.week_of_year(2017, 12, 31)
+      {2017, 52, 7}
+      iex> Calendar.ISO.week_of_year(2018, 1, 1)
+      {2018, 1, 1}
+      iex> Calendar.ISO.week_of_year(2018, 1, 2)
+      {2018, 1, 2}
+
+      # First day of the year is a tuesday
+      iex> Calendar.ISO.week_of_year(2018, 12, 31)
+      {2019, 1, 1}
+      iex> Calendar.ISO.week_of_year(2019, 1, 1)
+      {2019, 1, 2}
+      iex> Calendar.ISO.week_of_year(2019, 1, 2)
+      {2019, 1, 3}
+
+      # First day of the year is a wednesday
+      iex> Calendar.ISO.week_of_year(2013, 12, 29)
+      {2013, 52, 7}
+      iex> Calendar.ISO.week_of_year(2013, 12, 30)
+      {2014, 1, 1}
+      iex> Calendar.ISO.week_of_year(2013, 12, 31)
+      {2014, 1, 2}
+      iex> Calendar.ISO.week_of_year(2014, 1, 1)
+      {2014, 1, 3}
+      iex> Calendar.ISO.week_of_year(2014, 1, 2)
+      {2014, 1, 4}
+
+      # First day of the year is a thursday
+      iex> Calendar.ISO.week_of_year(2014, 12, 28)
+      {2014, 52, 7}
+      iex> Calendar.ISO.week_of_year(2014, 12, 29)
+      {2015, 1, 1}
+      iex> Calendar.ISO.week_of_year(2014, 12, 30)
+      {2015, 1, 2}
+      iex> Calendar.ISO.week_of_year(2014, 12, 31)
+      {2015, 1, 3}
+      iex> Calendar.ISO.week_of_year(2015, 1, 1)
+      {2015, 1, 4}
+      iex> Calendar.ISO.week_of_year(2015, 1, 2)
+      {2015, 1, 5}
+
+      # First day of the year is a friday
+      iex> Calendar.ISO.week_of_year(2016, 1, 1)
+      {2015, 53, 5}
+      iex> Calendar.ISO.week_of_year(2016, 1, 2)
+      {2015, 53, 6}
+      iex> Calendar.ISO.week_of_year(2016, 1, 3)
+      {2015, 53, 7}
+      iex> Calendar.ISO.week_of_year(2016, 1, 4)
+      {2016, 1, 1}
+
+      # First day of the year is a saturday
+      iex> Calendar.ISO.week_of_year(2011, 1, 1)
+      {2010, 52, 6}
+      iex> Calendar.ISO.week_of_year(2011, 1, 2)
+      {2010, 52, 7}
+      iex> Calendar.ISO.week_of_year(2011, 1, 3)
+      {2011, 1, 1}
+
+      # First day of the year is a sunday
+      iex> Calendar.ISO.week_of_year(2017, 1, 1)
+      {2016, 52, 7}
+      iex> Calendar.ISO.week_of_year(2017, 1, 2)
+      {2017, 1, 1}
+
+  Any other date:
+
+      iex> Calendar.ISO.week_of_year(2018, 12, 4)
+      {2018, 49, 2}
+
+  """
+  @doc since: "1.8.0"
+  @spec week_of_year(year, month, day) :: {year, 1..54, 1..7}
+  @impl true
+  def week_of_year(year, month, day) do
+    true = day <= days_in_month(year, month)
```

Comment:
Should this raise an `ArgumentError` instead of a `MatchError`?  I would also change `day_of_year` and all the functions that rely on `date_to_iso_days` to raise an `ArgumentError`.

Reply:
Sure, can you please send a PR after this one is merged? Thank you! :heart: